### ISSUE TITLE
Prevent users from staying in instance after disconnect

### ISSUE THREAD

**Issue by ldworkin** [HarvardEconCS/turkserver-meteor#27]
If a user disconnects in the middle of an instance, their record in Partitioner's Grouping collection stays around, and if they ever try to reconnect, "User already in group" errors get thrown.

**Comment by mizzao** [comment]
While fixing this, we should probably make sure that the accounting for disconnect time happens regardless of user is sent to lobby after teardown.

**Comment by ldworkin** [comment]
This bb6383a161470b0e5eba6d226e0cb28cc933473a should fix the main problem, though I don't know about the accounting for disconnect part.

**Comment by mizzao** [comment]
Yeah, this approach skips all of the accounting normally associated with removing a user from an instance, so I'd like to figure out how to preserve that while making sure that users don't experience this error.

**Comment by ldworkin** [comment]
Would calling `asst._leaveInstance` (in addition to `clearUserGroup`) fix this? I think this does the accounting you're referring to.

**Comment by ldworkin** [comment]
What if we modified _leaveInstance so that `now` is set as follows (and name of the variable changed accordingly) 1) if instance is not ended, this should be current time, and 2) if instance is ended, it should be instance end time. Would this work?

**Comment by mizzao** [comment]
Yeah, that seems good. I think it would be useful, in the future, to be able to support removing users from instances even before they are torn down.

**Comment by ldworkin** [comment]
We do support removing users from instances before they are torn down, right? Can't you call sendUserToLobby even if an instance is still going on?

And wrt your second point, yes, absolutely -- while working on my tutorial I drew up figures to illustrate the relationship between assignments, instances, users, hits, batches, etc; visuals like that would be helpful for both docs and a paper.

**Comment by mizzao** [comment]
@ldworkin you can call that, but I don't think I imagined using it in that way when I first wrote this, so we might need to consider any unintended side effects. I agree that it would be a great feature to have.

I think for a paper it would just be good to describe how the lobby/instances setup allows for different types of experiments, e.g.:

- Everyone does a tutorial separately then randomizes into a group (crisis mapping)
- Everyone joins small groups which are then re-matched over several supergames (cooperation)
- Several long-running worlds, running in parallel, have users joining and leaving them over a long period of time ("longitudinal")

For the last one we haven't actually done an experiment yet, but our cooperation experiment definitely proves that we can reduce attrition to a level where that is possible. Additionally, the longitudinal-and-synchronous nature of the cooperation experiment is a pretty powerful proof of concept that we should definitely highlight.

**Comment by ldworkin** [comment]
Just going back to the original issue for a sec, see here: cc383e9d9908589b6702178ff1991eeeec6937bc

Tested a bit and think it works.

**Comment by mizzao** [comment]
Looks great to me. I'd like to add some tests on top of that change, if that's okay with you.

Do you want to rebase `experimental` on `master` at your convenience, and then I'll just add some tests and/or other tweaks, and merge into master?

**Comment by ldworkin** [comment]
I have finished all changes I wanted to make for now and did the rebase, so test/tweak to your heart's content and merge away!
